src/librustc/ty/print/pretty.rs

```diff
@@ -969,69 +987,131 @@ pub trait PrettyPrinter<'tcx>:
+            // Char
+            (Scalar::Raw { data, .. }, ty::Char) =>
                 p!(write("{:?}", ::std::char::from_u32(data as u32).unwrap())),
-            (ConstValue::Scalar(_), ty::RawPtr(_)) => p!(write("{{pointer}}")),
-            (ConstValue::Scalar(Scalar::Ptr(ptr)), ty::FnPtr(_)) => {
+            // References and pointers
+            (Scalar::Raw { data: 0, .. }, ty::RawPtr(_)) => p!(write("{{null pointer}}")),
+            // This is UB, but we still print it
+            (Scalar::Raw { data: 0, ..}, ty::Ref(_, ty, _))
+                => p!(write("{{null reference to "), print(ty), write("}}")),
+            (Scalar::Raw { data, .. }, ty::Ref(..)) |
+            (Scalar::Raw { data, .. }, ty::RawPtr(_)) => {
+                let pointer_width = self.tcx().data_layout.pointer_size.bytes();
+                p!(write("0x{:01$x}", data, pointer_width as usize * 2))
+            },
+            (Scalar::Ptr(p), ty::Ref(..)) |
+            (Scalar::Ptr(p), ty::RawPtr(_)) => {
+                self = self.pretty_print_const_pointer(p)?
+            },
+            (Scalar::Ptr(ptr), ty::FnPtr(_)) => {
                 let instance = {
                     let alloc_map = self.tcx().alloc_map.lock();
                     alloc_map.unwrap_fn(ptr.alloc_id)
                 };
                 p!(print_value_path(instance.def_id(), instance.substs));
             },
-            _ => {
-                let printed = if let ty::Ref(_, ref_ty, _) = ty.kind {
-                    let byte_str = match (ct, &ref_ty.kind) {
-                        (ConstValue::Scalar(Scalar::Ptr(ptr)), ty::Array(t, n)) if *t == u8 => {
-                            let n = n.eval_usize(self.tcx(), ty::ParamEnv::empty());
-                            Some(self.tcx()
-                                .alloc_map.lock()
-                                .unwrap_memory(ptr.alloc_id)
-                                .get_bytes(&self.tcx(), ptr, Size::from_bytes(n)).unwrap())
-                        },
-                        (ConstValue::Slice { data, start, end }, ty::Slice(t)) if *t == u8 => {
-                            // The `inspect` here is okay since we checked the bounds, and there are
-                            // no relocations (we have an active slice reference here). We don't use
-                            // this result to affect interpreter execution.
-                            Some(data.inspect_with_undef_and_ptr_outside_interpreter(start..end))
-                        },
-                        _ => None,
-                    };
+            // For zsts just print their type as their value gives no extra information
+            (Scalar::Raw { size: 0, .. }, _) => p!(print(ty)),
+            // Nontrivial types with scalar bit representation
+            (Scalar::Raw { data, size }, _) => p!(
+                write("{{0x{:01$x} : ", data, size as usize * 2),
+                print(ty),
+                write("}}")
+            ),
```

Comment:
The ZST one should print like a value but the second one shouldn't. I'm actually surprised `print(ty)` makes it print like a value, you probably need to wrap in something the types that show up after `:`.

Reply:
Yea, the previous code did `write("{}", ty)` instead

Reply:
You should never use `write` to print things where `print` works, FWIW.

The fix here is to use one of the APIs that... wait, there isn't one right now ugh.

You'll have to add something like this: https://github.com/rust-lang/rust/blob/8960acf023d6739a601ae83cfc3d4ac30a85893e/src/librustc/ty/print/pretty.rs#L1362-L1374

But instead of printing `<T>` it prints `: T`. Or maybe you can make it so the value is printed through it too (i.e. use two closures) so you can get `{x: T}` out of it in one go (you could call it `type_ascribed` or something).